rest/endpoints/license/user_licenses_v4.yml

```diff
@@ -0,0 +1,100 @@
+        - in: path
+          name: user_id
+          required: true
+          description: "The user identifier"
+          schema:
+            type: string
+        - in: header
+          name: Idempotency-Key
+          required: true
+          description: "Client unique generated value for server to recognize subsequent retries"
+          schema:
+            type: string
+          example: "518d5812-b8e8-43a4-b3b0-8f9c5cdc31b9"
+      requestBody:
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: "../../snippets/Config/LicenseEntity.yml#/CreateLicenseEntity"
+
+      responses:
+        201:
+          description: "Successful response"
+          content:
+            application/json:
+              schema:
+                type: object
+                required:
+                - data
+                properties:
+                  data:
+                    $ref: "https://raw.githubusercontent.com/booxmedialtd/api-documentation/master/rest/snippets/Config/LicenseEntity.yml#/LicenseEntity"
+        403:
+          description: Incorrect PIN code
+          content:
+            application/json:
+              schema:
+                $ref: "https://raw.githubusercontent.com/booxmedialtd/api-documentation/master/rest/snippets/General/JsonApiErrorEntity.yml#/JsonApiErrorEntity"
+              example:
+                errors:
+                  - id: "2d3ba0ea-0aca-11ea-b2d0-4bc6fcad2537"
+                    status: "403"
+                    code: "10200"
+                    title: "Incorrect PIN code"
+                    detail: 'The PIN code that you supplied does not match the one we have stored for the account.'
+        404:
+          description: Not Found
+          content:
+            application/json:
+              schema:
+                $ref: "https://raw.githubusercontent.com/booxmedialtd/api-documentation/master/rest/snippets/General/JsonApiErrorEntity.yml#/JsonApiErrorEntity"
+              example:
+                errors:
+                  - id: "2d3ba0ea-0aca-11ea-b2d0-4bc6fcad2537"
+                    status: "404"
+                    code: "30701"
+                    title: "Unable to purchase license"
+                    detail: 'Product "5000" not found'
+        409:
+          description: Conflict
+          content:
+            application/json:
+              schema:
+                $ref: "https://raw.githubusercontent.com/booxmedialtd/api-documentation/master/rest/snippets/General/JsonApiErrorEntity.yml#/JsonApiErrorEntity"
+              example:
+                errors:
+                  - id: "2d3ba0ea-0aca-11ea-b2d0-4bc6fcad2537"
+                    status: "409"
+                    code: "30703"
+                    title: "Already processed license"
+                    detail: 'The user "1200000" license (with idempotency key "518d5812-b8e8-43a4-b3b0-8f9c5cdc31b9") has been processed, or in progress'
```

Comment:
This behaviour is wrong.
According to the implementation guide I linked: https://multithreaded.stitchfix.com/blog/2017/06/26/patterns-of-soa-idempotency-key/
the API must respond with the same response as it did the first time the request was made. Otherwise, the client will have the following problem:
If the request is processed but the response is lost due to some network error, then the client has no way of knowing. They will attempt the request again. This time the API responds with this 409 error. The client knows the order was processed the first round but they still need the data returned from the original request to proceed (in this case the License object, possibly to display it to the user), but has no way of obtaining it.

Reply:
changed